frontend/src/components/pages/roles/RoleUpdatePage.tsx

```diff
@@ -0,0 +1,214 @@
+ * @param newRules - The desired ACL rules
+ * @returns Object containing arrays of rules to create and delete
+ */
+function calculateACLDifference(currentRules: ACLWithId[], newRules: ACLWithId[]): ACLDifference {
+  const currentIds = new Set(currentRules.map((r) => r.id));
+  const newIds = new Set(newRules.map((r) => r.id));
+
+  // Rules to create: in newRules but not in currentRules
+  const toCreate = newRules.filter((rule) => !currentIds.has(rule.id));
+
+  // Rules to delete: in currentRules but not in newRules
+  const toDelete = currentRules.filter((rule) => !newIds.has(rule.id));
+
+  return {
+    toCreate,
+    toDelete,
+  };
+}
+
+const RoleUpdatePage = () => {
+  const toast = useToast();
+  const nav = useNavigate();
+  const queryClient = useQueryClient();
+  const { roleName = '' } = useParams<{ roleName: string }>();
+
+  useEffect(() => {
+    uiState.pageBreadcrumbs = [
+      { title: 'Security', linkTo: '/security' },
+      { title: 'Roles', linkTo: '/security/roles' },
+      { title: roleName, linkTo: `/security/roles/${roleName}/details` },
+      { title: 'Update Role', linkTo: '', heading: '' },
+    ];
+  }, [roleName]);
+
+  // Fetch existing ACL data for the role
+  const { data, isLoading } = useQuery(
+    listACLs,
+    {
+      filter: {
+        principal: `RedpandaRole:${roleName}`,
+      },
+    } as ListACLsRequest,
+    {
+      enabled: !!roleName,
+      select: getAclFromAclListResponse,
+    },
+  );
+
+  const { mutateAsync: createACLMutation } = useMutation(createACL);
+  const { mutateAsync: deleteACLMutation } = useMutation(deleteACLs);
+
+  const updateRoleAclMutation =
+    (actualRules: Rule[], sharedConfig: SharedConfig) => async (_: string, _2: string, rules: Rule[]) => {
+      const currentRules: ACLWithId[] = convertRulesToCreateACLRequests(
+        actualRules,
+        sharedConfig.principal,
+        sharedConfig.host,
+      ).map((r) => ({
+        ...r,
+        id: getIdFromCreateACLRequest(r),
+      }));
+      const newRules: ACLWithId[] = convertRulesToCreateACLRequests(
+        rules,
+        sharedConfig.principal,
+        sharedConfig.host,
+      ).map((r) => ({
+        ...r,
+        id: getIdFromCreateACLRequest(r),
+      }));
+
+      const { toCreate, toDelete } = calculateACLDifference(currentRules, newRules);
+
+      console.log('Rules to create:', toCreate);
+      console.log('Rules to delete:', toDelete);
+
+      const createResults = toCreate.map((r) => createACLMutation(r));
+      const deleteResults = toDelete.map((r) => {
+        const d: DeleteACLsRequest = {
+          filter: {
+            principal: r.principal,
+            resourceType: r.resourceType,
+            resourceName: r.resourceName,
+            host: r.host,
+            operation: r.operation,
+            permissionType: r.permissionType,
+            resourcePatternType: r.resourcePatternType,
+            $typeName: 'redpanda.api.dataplane.v1.DeleteACLsRequest.Filter',
```

Comment:
You shouldn't need to do `$typeName` manually, just use `create()`

Reply:
done